File: Design/alog/first/数据结构/二叉树/ts/二叉树中第K小的元素.ts
import type { TreeNode } from './types'

function kthSmallest(root: TreeNode | null, k: number): number {
  let cnt = 0;
  let result = 0;

  const dfs = (node: TreeNode | null): void => {
    if (node === null) {
      return;
    }

    dfs(node.left);

    cnt++;
    if (cnt === k) {
      result = node.val;
      return;
    }

    dfs(node.right);
  }

  dfs(root);

  return result;
};

File: Design/alog/first/数据结构/二叉树/py/二叉树中第K小的元素.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        node_val = 0
        cnt = 0

        def inorder(node: TreeNode) -> None:
            nonlocal node_val, cnt

            if not node:
                return

            inorder(node.left)
            cnt += 1
            if cnt == k:
                node_val = node.val
                return
            inorder(node.right)

        inorder(root)

        return node_val
